actions/restore.py

```diff
@@ -211,10 +212,46 @@ def shasum_file(filepath):
     return shasum.hexdigest()
 
 
+def dismantle_cluster():
+    """Disconnect other cluster members.
+
+    This is a preparation step before restoring snapshot on the cluster.
+    """
+    log('Disconnecting cluster members')
+    etcdctl = EtcdCtl()
+    etcd_conf = EtcdDatabag()
+
+    my_name = etcd_conf.unit_name
+    endpoint = 'https://{}:{}'.format(etcd_conf.cluster_address,
+                                      etcd_conf.port)
+    for name, data in etcdctl.member_list(endpoint).items():
+        if name != my_name:
+            log('Disconnecting {}'.format(name), hookenv.DEBUG)
+            etcdctl.unregister(data['unit_id'], endpoint)
+
+    etcd_conf.cluster_state = 'new'
+    conf_path = os.path.join(etcd_conf.etcd_conf_dir, "etcd.conf.yml")
+    render('etcd3.conf', conf_path, etcd_conf.__dict__, owner='root',
+           group='root')
+
+
+def rebuild_cluster():
+    """Signal other etcd units to rejoin new cluster."""
+    log('Requesting peer members to rejoin cluster')
+    peer_rel_id = hookenv.peer_relation_id()
+    rejoin_request = uuid4().hex
+    hookenv.relation_set(peer_rel_id, force_rejoin=rejoin_request)
```

Comment:
I think it's also the case that the leader unit wouldn't even get a `cluster-relation-changed` hook for the `force_rejoin` flag, since units don't get notified of updates to their own data.  That's also true, however, for leadership data, and additionally the handlers aren't invoked following an action, so it would depend on another hook invocation for the leader regardless.

My suggestion is to move the `force_rejoin` and `check_cluster_health` functions to `lib/etcd_lib.py` and import and call them directly from the action, as well as importing and calling them from the handlers.  (You'd have to keep a handler for `check_cluster_health` with the same decorators, but it could just call out to the function imported from `etcd_lib`; though you'd have to rename one of them to avoid a conflict.)

Reply:
> I think it's also the case that the leader unit wouldn't even get a cluster-relation-changed hook for the force_rejoin flag

That's not a n issue here as the leader unit starts new cluster itself and needs only the other units to rejoin. Other than that, I'll gent on the suggested changes.